src/backend/tty.rs

```diff
@@ -1102,13 +1104,114 @@ impl Tty {
             }
         };
 
+        let overlay_cursor = niri
+            .output_state
+            .get(output)
+            .and_then(|state| state.pending_screencopy.as_ref())
+            .map_or(true, |s| s.overlay_cursor());
+
         // Render the elements.
-        let elements = niri.render::<TtyRenderer>(&mut renderer, output, true);
+        let elements = niri.render::<TtyRenderer>(&mut renderer, output, overlay_cursor);
 
         // Hand them over to the DRM.
         let drm_compositor = &mut surface.compositor;
         match drm_compositor.render_frame::<_, _>(&mut renderer, &elements, [0.; 4]) {
             Ok(res) => {
+                if let Some(screencopy) = niri
+                    .output_state
+                    .get_mut(output)
+                    .and_then(|s| s.pending_screencopy.take())
+                {
+                    let region = screencopy.region();
+                    let damage = [Rectangle::from_loc_and_size((0, 0), region.size)];
+                    let blit: anyhow::Result<_> = if let Ok(dmabuf) =
+                        dmabuf::get_dmabuf(screencopy.buffer())
+                    {
+                        renderer.bind(dmabuf).unwrap();
+
+                        res.blit_frame_result(
+                            screencopy.region().size,
+                            Transform::Normal,
+                            output.current_scale().fractional_scale(),
+                            &mut renderer,
+                            damage,
+                            [],
+                        )
+                        .map(|sync| sync.export())
+                        .map_err(Into::into)
+                    } else {
+                        let renderer = renderer.as_gles_renderer();
+                        shm::with_buffer_contents_mut(
+                            screencopy.buffer(),
+                            |shm_buffer, shm_len, buffer_data| {
+                                let region = screencopy.region();
+                                if !(buffer_data.format == wl_shm::Format::Argb8888
+                                    && buffer_data.stride == region.size.w * 4
+                                    && buffer_data.height == region.size.h
+                                    && shm_len as i32 == buffer_data.stride * buffer_data.height)
+                                {
+                                    return Err(anyhow!("the buffer format is incorrect"));
+                                }
+
+                                renderer.with_context(|gl| {
+                                    // SAFETY: i hope this is okay
+                                    unsafe {
+                                        gl.ReadPixels(
```

Comment:
I think you can use a combination of the same as what `render_helpers::render_to_vec()` does and rendering to a small texture directly with the right `RelocateRenderElement` to put the target square into it.

Reply:
Do you mean, we render to a `Vec<u8>` and then copy those bytes? Is there some safe function to map a texture onto a `wl_shm` buffer? I don't understand why else we would do this, and i don't see a significant issue with the current way of doing it (besides that comment of course, but that was written in my "blindly copy and paste" stage - i feel that i understand a bit better now)

Reply:
We always render to a GPU texture. Then `render_to_vec()` maps that GPU texture to the main memory and copies it into a Vec; you can copy into the shm WlBuffer instead.

Reply:
Isn't that essentially what i'm already doing, just from a different buffer? As far as i can tell, i think what you're describing is just what `ReadPixels` does? I don't think i fully understand the suggestion.